File: src/store/offlineStore.js
import { generateSchedule, assessBorrower } from '../utils/loanUtils';

const STORAGE_PREFIX = 'microlend_';
const QUEUE_KEY = `${STORAGE_PREFIX}write_queue`;

// Event listeners for store reactivity
const listeners = new Set();

// In-memory cache to maintain reference stability for useSyncExternalStore
let cache = {};

export function clearStoreCache() {
  cache = {};
}

function notify() {
  listeners.forEach((listener) => listener());
}

/**
 * Gets item from LocalStorage with in-memory caching for reference stability.
 */
function getItem(key, defaultValue = null) {
  if (cache[key] !== undefined) {
    return cache[key];
  }
  try {
    const data = localStorage.getItem(key);
    const parsed = data ? JSON.parse(data) : defaultValue;
    cache[key] = parsed;
    return parsed;
  } catch (e) {
    console.error(`Error reading ${key} from LocalStorage`, e);
    cache[key] = defaultValue;
    return defaultValue;
  }
}

/**
 * Sets raw item in LocalStorage and updates cache.
 */
function setItem(key, value) {
  try {
    cache[key] = value;
    localStorage.setItem(key, JSON.stringify(value));
  } catch (e) {
    console.error(`Error writing ${key} to LocalStorage`, e);
  }
}

/**
 * Write queue operations
 */
export function getQueue() {
  return getItem(QUEUE_KEY, []);
}

function pushToQueue(operation) {
  const queue = getQueue();
  const updatedQueue = [
    ...queue,
    {
      id: `queue_${Date.now()}_${Math.random().toString(36).substr(2, 5)}`,
      timestamp: new Date().toISOString(),
      ...operation,
    },
  ];
  setItem(QUEUE_KEY, updatedQueue);
}

/**
 * Sync all queued write actions.
 */
export function syncAll() {
  return new Promise((resolve) => {
    setTimeout(() => {
      setItem(QUEUE_KEY, []);
      notify();
      resolve({ success: true, syncedCount: getQueue().length });
    }, 500);
  });
}

/**
 * Collection operations
 */
export function getCollection(collectionName) {
  return getItem(`${STORAGE_PREFIX}${collectionName}`, []);
}

export function saveCollection(collectionName, items) {
  setItem(`${STORAGE_PREFIX}${collectionName}`, items);
  notify();
}

export function addItem(collectionName, item) {
  const collection = getCollection(collectionName);
  const newItem = {
    id: item.id || `${collectionName.slice(0, 3)}_${Date.now()}_${Math.random().toString(36).substr(2, 4)}`,
    createdAt: new Date().toISOString(),
    ...item,
  };
  const updated = [newItem, ...collection];
  saveCollection(collectionName, updated);

  pushToQueue({
    action: 'CREATE',
    collection: collectionName,
    payload: newItem,
  });

  return newItem;
}

export function updateItem(collectionName, id, updates) {
  const collection = getCollection(collectionName);
  let updatedItem = null;

  const updated = collection.map((item) => {
    if (item.id === id) {
      updatedItem = { ...item, ...updates, updatedAt: new Date().toISOString() };
      return updatedItem;
    }
    return item;
  });

  saveCollection(collectionName, updated);

  if (updatedItem) {
    pushToQueue({
      action: 'UPDATE',
      collection: collectionName,
      id,
      payload: updates,
    });
  }

  return updatedItem;
}

export function deleteItem(collectionName, id) {
  const collection = getCollection(collectionName);
  const updated = collection.filter((item) => item.id !== id);
  saveCollection(collectionName, updated);

  pushToQueue({
    action: 'DELETE',
    collection: collectionName,
    id,
  });
}

/**
 * Pub/Sub subscription for store changes
 */
export function subscribe(listener) {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

/**
 * Seed initial sample data if store is empty.
 */
export function seedInitialData() {
  const existingBorrowers = getCollection('borrowers');
  const existingLoans = getCollection('loans');

  if (existingBorrowers.length > 0 || existingLoans.length > 0) {
    return;
  }

  const today = new Date();

  const monthsAgo = (m) => {
    const d = new Date(today);
    d.setMonth(d.getMonth() - m);
    return d.toISOString().split('T')[0];
  };

  const sampleBorrowers = [
    {
      id: 'bor_elena_01',
      full_name: 'Elena Rostova',
      email: 'elena.rostova@example.com',
      phone: '+1 (555) 234-5678',
      address: '742 Evergreen Terrace, Springfield, IL',
      id_number: 'ID-8839201',
      employment: 'Senior Software Engineer at TechCorp',
      monthly_income: 6800,
      credit_score: 82,
      risk_rating: 'low',
      notes: 'Reliable client, always pays on or ahead of time. High income-to-debt ratio.',
      createdAt: monthsAgo(6),
    },
    {
      id: 'bor_marcus_02',
      full_name: 'Marcus Vance',
      email: 'm.vance@example.com',
      phone: '+1 (555) 876-5432',
      address: '1048 Ocean Avenue, Santa Monica, CA',
      id_number: 'ID-4421098',
      employment: 'Retail Store Manager at Urban Outfitters',
      monthly_income: 3900,
      credit_score: 62,
      risk_rating: 'medium',
      notes: 'Seasonal store sales fluctuations. Keeps good communication.',
      createdAt: monthsAgo(4),
    },
    {
      id: 'bor_aisha_03',
      full_name: 'Aisha Patel',
      email: 'aisha.design@example.com',
      phone: '+1 (555) 345-6789',
      address: '302 Pine Street, Austin, TX',
      id_number: 'ID-9920112',
      employment: 'Freelance UI/UX Designer',
      monthly_income: 2900,
      credit_score: 55,
      risk_rating: 'medium',
      notes: 'New applicant requesting working capital for new workstation setup.',
      createdAt: monthsAgo(1),
    },
  ];

  saveCollection('borrowers', sampleBorrowers);

  // Loan 1: Elena Rostova - Active, 12 months, $5000 @ 10%, disbursed 2 months ago
  const loan1Disbursed = monthsAgo(2);
  const loan1Schedule = generateSchedule(5000, 10, 12, loan1Disbursed);

  const loan1Payments = [
    {
      id: 'pay_01',
      date: monthsAgo(1),
      amount: loan1Schedule[0].amount,
      method: 'Bank Transfer',
      note: 'Installment 1 paid',
    },
    {
      id: 'pay_02',
      date: monthsAgo(0),
      amount: loan1Schedule[1].amount,
      method: 'Bank Transfer',
      note: 'Installment 2 paid',
    },
  ];

  // Loan 2: Marcus Vance - Active with overdue payment
  const loan2Disbursed = monthsAgo(3);
  const loan2Schedule = generateSchedule(3000, 14, 6, loan2Disbursed);

  const loan2Payments = [
    {
      id: 'pay_03',
      date: monthsAgo(2),
      amount: loan2Schedule[0].amount,
      method: 'Debit Card',
      note: 'Installment 1 paid',
    },
    {
      id: 'pay_04',
      date: monthsAgo(1),
      amount: 150,
      method: 'Cash',
      note: 'Partial payment for installment 2',
    },
  ];

  // Loan 3: Aisha Patel - Pending loan
  const loan3Disbursed = today.toISOString().split('T')[0];
  const loan3Schedule = generateSchedule(2500, 12, 6, loan3Disbursed);

  const sampleLoans = [
    {
      id: 'loan_elena_01',
      borrower_id: 'bor_elena_01',
      principal: 5000,
      interest_rate: 10,
      term_months: 12,
      purpose: 'Tech Equipment Purchase',
      status: 'active',
      disbursement_date: loan1Disbursed,
      credit_assessment: assessBorrower(sampleBorrowers[0]),
      schedule: loan1Schedule,
      payments: loan1Payments,
      notes: 'Computer upgrade financing for freelance business expansion.',
      createdAt: loan1Disbursed,
    },
    {
      id: 'loan_marcus_02',
      borrower_id: 'bor_marcus_02',
      principal: 3000,
      interest_rate: 14,
      term_months: 6,
      purpose: 'Retail Inventory Restock',
      status: 'active',
      disbursement_date: loan2Disbursed,
      credit_assessment: assessBorrower(sampleBorrowers[1]),
      schedule: loan2Schedule,
      payments: loan2Payments,
      notes: 'Inventory financing for seasonal merchandise.',
      createdAt: loan2Disbursed,
    },
    {
      id: 'loan_aisha_03',
      borrower_id: 'bor_aisha_03',
      principal: 2500,
      interest_rate: 12,
      term_months: 6,
      purpose: 'Studio Design Workstation',
      status: 'pending',
      disbursement_date: loan3Disbursed,
      credit_assessment: assessBorrower(sampleBorrowers[2]),
      schedule: loan3Schedule,
      payments: [],
      notes: 'Awaiting final approval on design portfolio verification.',
      createdAt: today.toISOString().split('T')[0],
    },
  ];

  saveCollection('loans', sampleLoans);
}

File: flutter_app/lib/store/offline_store.dart
import 'dart:async';
import 'dart:convert';
import 'package:intl/intl.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/borrower.dart';
import '../models/loan.dart';
import '../models/payment.dart';
import '../models/user.dart';
import '../utils/loan_utils.dart';

class OfflineStore {
  Future<void> _lock = Future.value();

  Future<T> _synchronized<T>(Future<T> Function() operation) async {
    final previous = _lock;
    final completer = Completer<void>();
    _lock = completer.future;
    await previous;
    try {
      return await operation();
    } finally {
      completer.complete();
    }
  }
  static const String _storagePrefix = 'microlend_';
  static const String _settingPrefix = '${_storagePrefix}setting_';

  final SharedPreferences _prefs;

  OfflineStore(this._prefs);

  static Future<OfflineStore> init() async {
    final prefs = await SharedPreferences.getInstance();
    final store = OfflineStore(prefs);
    await store.seedInitialData();
    return store;
  }

  Future<void> reload() async {
    try {
      await _prefs.reload();
    } catch (_) {}
  }

  String getSetting(String key, String defaultValue) {
    return _prefs.getString('$_settingPrefix$key') ?? defaultValue;
  }

  Future<void> setSetting(String key, String value) async {
    await _synchronized(() async {
      await _prefs.setString('$_settingPrefix$key', value);
    });
  }

  List<Map<String, dynamic>> getCollection(String collectionName) {
    final key = '$_storagePrefix$collectionName';
    final raw = _prefs.getString(key);
    if (raw == null || raw.isEmpty) return [];
    try {
      final List decoded = jsonDecode(raw);
      return decoded.map((e) => Map<String, dynamic>.from(e)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<void> saveCollection(String collectionName, List<Map<String, dynamic>> items) async {
    await _synchronized(() async {
      final key = '$_storagePrefix$collectionName';
      await _prefs.setString(key, jsonEncode(items));
    });
  }

  Future<Map<String, dynamic>> addItem(String collectionName, Map<String, dynamic> item) async {
    return await _synchronized(() async {
      final key = '$_storagePrefix$collectionName';
      final raw = _prefs.getString(key);
      List<Map<String, dynamic>> collection = [];
      if (raw != null && raw.isNotEmpty) {
        try {
          final List decoded = jsonDecode(raw);
          collection = decoded.map((e) => Map<String, dynamic>.from(e)).toList();
        } catch (_) {}
      }

      final newItem = {
        'id': item['id'] ?? '${collectionName.substring(0, 3)}_${DateTime.now().millisecondsSinceEpoch}',
        'createdAt': DateTime.now().toIso8601String(),
        ...item,
      };

      final updated = [newItem, ...collection];
      await _prefs.setString(key, jsonEncode(updated));
      return newItem;
    });
  }

  Future<Map<String, dynamic>?> updateItem(String collectionName, String id, Map<String, dynamic> updates) async {
    return await _synchronized(() async {
      final key = '$_storagePrefix$collectionName';
      final raw = _prefs.getString(key);
      List<Map<String, dynamic>> collection = [];
      if (raw != null && raw.isNotEmpty) {
        try {
          final List decoded = jsonDecode(raw);
          collection = decoded.map((e) => Map<String, dynamic>.from(e)).toList();
        } catch (_) {}
      }

      Map<String, dynamic>? updatedItem;
      final updated = collection.map((item) {
        if (item['id'] == id) {
          updatedItem = {
            ...item,
            ...updates,
            'updatedAt': DateTime.now().toIso8601String(),
          };
          return updatedItem!;
        }
        return item;
      }).toList();

      await _prefs.setString(key, jsonEncode(updated));
      return updatedItem;
    });
  }

  Future<void> deleteItem(String collectionName, String id) async {
    await _synchronized(() async {
      final key = '$_storagePrefix$collectionName';
      final raw = _prefs.getString(key);
      List<Map<String, dynamic>> collection = [];
      if (raw != null && raw.isNotEmpty) {
        try {
          final List decoded = jsonDecode(raw);
          collection = decoded.map((e) => Map<String, dynamic>.from(e)).toList();
        } catch (_) {}
      }
      final updated = collection.where((item) => item['id'] != id).toList();
      await _prefs.setString(key, jsonEncode(updated));
    });
  }

  Future<Map<String, dynamic>?> appendToItemArray(
    String collectionName,
    String id,
    String arrayField,
    Map<String, dynamic> element, {
    String idField = 'id',
  }) async {
    return await _synchronized(() async {
      final key = '$_storagePrefix$collectionName';
      final raw = _prefs.getString(key);
      List<Map<String, dynamic>> collection = [];
      if (raw != null && raw.isNotEmpty) {
        try {
          final List decoded = jsonDecode(raw);
          collection = decoded.map((e) => Map<String, dynamic>.from(e)).toList();
        } catch (_) {}
      }

      Map<String, dynamic>? updatedItem;
      final updated = collection.map((item) {
        if (item['id'] == id) {
          final existingArrayRaw = item[arrayField] is List ? List.from(item[arrayField]) : [];
          final existingArray = existingArrayRaw.map((e) => Map<String, dynamic>.from(e as Map)).toList();

          final elementId = element[idField];
          final exists = elementId != null && existingArray.any((e) => e[idField] == elementId);

          if (!exists) {
            existingArray.add(element);
          }

          updatedItem = {
            ...item,
            arrayField: existingArray,
            'updatedAt': DateTime.now().toIso8601String(),
          };
          return updatedItem!;
        }
        return item;
      }).toList();

      await _prefs.setString(key, jsonEncode(updated));
      return updatedItem;
    });
  }

  Future<void> clearAllData() async {
    await saveCollection('borrowers', []);
    await saveCollection('loans', []);
    await saveCollection('users', []);
  }

  Future<void> seedInitialData({bool force = false}) async {
    final existingBorrowers = getCollection('borrowers');
    final existingLoans = getCollection('loans');
    final existingUsers = getCollection('users');

    if (force || existingUsers.isEmpty) {
      final adminSalt = User.generateSalt();
      final defaultAdmin = User(
        id: 'usr_admin',
        username: 'admin',
        passwordHash: User.hashPassword('admin123', adminSalt),
        salt: adminSalt,
        role: 'approver',
        mustChangePassword: true,
      );
      await saveCollection('users', [defaultAdmin.toMap()]);
    }

    if (!force && (existingBorrowers.isNotEmpty || existingLoans.isNotEmpty)) {
      return;
    }

    if (force) {
      await clearAllData();
    }

    final today = DateTime.now();

    String monthsAgo(int m) {
      final d = DateTime(today.year, today.month - m, today.day);
      return DateFormat('yyyy-MM-dd').format(d);
    }

    final sampleBorrowers = [
      Borrower(
        id: 'bor_elena_01',
        fullName: 'Elena Rostova',
        email: 'elena.rostova@example.com',
        phone: '+1 (555) 234-5678',
        address: '742 Evergreen Terrace, Springfield, IL',
        idNumber: 'ID-8839201',
        employment: 'Senior Software Engineer at TechCorp',
        monthlyIncome: 6800.0,
        creditScore: 82,
        riskRating: 'low',
        notes: 'Reliable client, always pays on or ahead of time. High income-to-debt ratio.',
        createdAt: monthsAgo(6),
      ),
      Borrower(
        id: 'bor_marcus_02',
        fullName: 'Marcus Vance',
        email: 'm.vance@example.com',
        phone: '+1 (555) 876-5432',
        address: '1048 Ocean Avenue, Santa Monica, CA',
        idNumber: 'ID-4421098',
        employment: 'Retail Store Manager at Urban Outfitters',
        monthlyIncome: 3900.0,
        creditScore: 62,
        riskRating: 'medium',
        notes: 'Seasonal store sales fluctuations. Keeps good communication.',
        createdAt: monthsAgo(4),
      ),
      Borrower(
        id: 'bor_aisha_03',
        fullName: 'Aisha Patel',
        email: 'aisha.design@example.com',
        phone: '+1 (555) 345-6789',
        address: '302 Pine Street, Austin, TX',
        idNumber: 'ID-9920112',
        employment: 'Freelance UI/UX Designer',
        monthlyIncome: 2900.0,
        creditScore: 55,
        riskRating: 'medium',
        notes: 'New applicant requesting working capital for new workstation setup.',
        createdAt: monthsAgo(1),
      ),
    ];

    await saveCollection('borrowers', sampleBorrowers.map((b) => b.toMap()).toList());

    // Loan 1
    final loan1Disbursed = monthsAgo(2);
    final loan1Schedule = LoanUtils.generateSchedule(5000.0, 10.0, 12, loan1Disbursed);
    final loan1Payments = [
      Payment(
        id: 'pay_01',
        date: monthsAgo(1),
        amount: loan1Schedule[0].amount,
        method: 'Bank Transfer',
        note: 'Installment 1 paid',
      ),
      Payment(
        id: 'pay_02',
        date: monthsAgo(0),
        amount: loan1Schedule[1].amount,
        method: 'Bank Transfer',
        note: 'Installment 2 paid',
      ),
    ];

    // Loan 2
    final loan2Disbursed = monthsAgo(3);
    final loan2Schedule = LoanUtils.generateSchedule(3000.0, 14.0, 6, loan2Disbursed);
    final loan2Payments = [
      Payment(
        id: 'pay_03',
        date: monthsAgo(2),
        amount: loan2Schedule[0].amount,
        method: 'Debit Card',
        note: 'Installment 1 paid',
      ),
      Payment(
        id: 'pay_04',
        date: monthsAgo(1),
        amount: 150.0,
        method: 'Cash',
        note: 'Partial payment for installment 2',
      ),
    ];

    // Loan 3
    final loan3Disbursed = DateFormat('yyyy-MM-dd').format(today);
    final loan3Schedule = LoanUtils.generateSchedule(2500.0, 12.0, 6, loan3Disbursed);

    final sampleLoans = [
      Loan(
        id: 'loan_elena_01',
        borrowerId: 'bor_elena_01',
        principal: 5000.0,
        interestRate: 10.0,
        termMonths: 12,
        purpose: 'Tech Equipment Purchase',
        status: 'active',
        disbursementDate: loan1Disbursed,
        creditAssessment: LoanUtils.assessBorrower(sampleBorrowers[0], []),
        schedule: loan1Schedule,
        payments: loan1Payments,
        notes: 'Computer upgrade financing for freelance business expansion.',
        createdAt: loan1Disbursed,
      ),
      Loan(
        id: 'loan_marcus_02',
        borrowerId: 'bor_marcus_02',
        principal: 3000.0,
        interestRate: 14.0,
        termMonths: 6,
        purpose: 'Retail Inventory Restock',
        status: 'active',
        disbursementDate: loan2Disbursed,
        creditAssessment: LoanUtils.assessBorrower(sampleBorrowers[1], []),
        schedule: loan2Schedule,
        payments: loan2Payments,
        notes: 'Inventory financing for seasonal merchandise.',
        createdAt: loan2Disbursed,
      ),
      Loan(
        id: 'loan_aisha_03',
        borrowerId: 'bor_aisha_03',
        principal: 2500.0,
        interestRate: 12.0,
        termMonths: 6,
        purpose: 'Studio Design Workstation',
        status: 'pending',
        disbursementDate: loan3Disbursed,
        creditAssessment: LoanUtils.assessBorrower(sampleBorrowers[2], []),
        schedule: loan3Schedule,
        payments: [],
        notes: 'Awaiting final approval on design portfolio verification.',
        createdAt: loan3Disbursed,
      ),
    ];

    await saveCollection('loans', sampleLoans.map((l) => l.toMap()).toList());
  }
}
